controllers/stargate_controller.go

```diff
@@ -0,0 +1,335 @@
+
+	// Check if a deployment already exists, if not create a new one
+	deploymentName := deploymentName(cassdc)
+	deploymentKey := client.ObjectKey{Namespace: req.Namespace, Name: deploymentName}
+	actualDeployment := &appsv1.Deployment{}
+	logger.Info("Fetching Stargate Deployment", "Deployment", deploymentKey)
+	if err := r.Get(ctx, deploymentKey, actualDeployment); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Stargate Deployment not found, creating a new one", "Deployment", deploymentKey)
+			if err := r.Create(ctx, desiredDeployment); err != nil {
+				logger.Error(err, "Failed to create new Stargate Deployment", "Deployment", deploymentKey)
+				return ctrl.Result{}, err
+			}
+			logger.Info("Stargate Deployment created successfully", "Deployment", deploymentKey)
+			return ctrl.Result{RequeueAfter: time.Minute}, nil
+		}
+		logger.Error(err, "Failed to fetch Stargate Deployment", "Deployment", deploymentKey)
+		return ctrl.Result{}, err
+	}
+
+	// Check if the deployment needs to be updated
+	desiredHash := deepHashString(desiredDeployment)
+	desiredDeployment.Annotations[resourceHashAnnotation] = desiredHash
+	if actualHash, found := actualDeployment.Annotations[resourceHashAnnotation]; !found || actualHash != desiredHash {
+		logger.Info("Updating Stargate Deployment", "Deployment", deploymentKey)
+		if err := r.Update(ctx, desiredDeployment); err != nil {
+			logger.Error(err, "Failed to update Stargate Deployment", "Deployment", deploymentKey)
+			return ctrl.Result{}, err
+		}
+		logger.Info("Stargate Deployment updated successfully", "Deployment", deploymentKey)
+		return ctrl.Result{RequeueAfter: time.Minute}, nil
+	}
+
+	//stargate.Status.DeploymentName = deploymentName
+	stargate.Status.Replicas = actualDeployment.Status.Replicas
+	stargate.Status.ReadyReplicas = actualDeployment.Status.ReadyReplicas
+	stargate.Status.UpdatedReplicas = actualDeployment.Status.UpdatedReplicas
+	stargate.Status.AvailableReplicas = actualDeployment.Status.AvailableReplicas
+	stargate.Status.ReadyStatus = fmt.Sprintf("%v/%v", actualDeployment.Status.ReadyReplicas, actualDeployment.Status.Replicas)
+
+	if err := r.Status().Update(ctx, stargate); err != nil {
+		logger.Error(err, "Failed to update Stargate status", "Stargate", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Stargate Deployment is up to date with Stargate resource", "Stargate", req.NamespacedName, "Deployment", deploymentKey)
+
+	return ctrl.Result{}, nil
+}
+
+func deploymentName(cassdc *cassdcapi.CassandraDatacenter) string {
+	// FIXME is this correct?
+	//  {{ .Release.Name }}-{{ include "k8ssandra.datacenterName" . }}-stargate
+	return cassdc.Spec.ClusterName + "-" + cassdc.Name + "-stargate"
+}
+
+// newStargateDeployment returns a Deployment object for Stargate
+func (r *StargateReconciler) newStargateDeployment(stargate *api.Stargate, cassdc *cassdcapi.CassandraDatacenter) *appsv1.Deployment {
```

Comment:
I would make `newStargateDeployment` a stand alone function instead of a method of StargateReconciler. This will make it a lot easier for writing unit tests.